File: app/src/main/java/no/simula/corona/data/greendao/BluetoothContact.java
package no.simula.corona.data.greendao;


import org.greenrobot.greendao.annotation.Entity;
import org.greenrobot.greendao.annotation.Id;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.TimeZone;

import no.simula.corona.Utils;
import org.greenrobot.greendao.annotation.Generated;

@Entity
public class BluetoothContact {

    @Id(autoincrement = true)
    private Long id;

    private Long timestamp;
    private String deviceId;
    private int rssi;
    private int txPower;
    private Boolean isUploaded = false;


    @Generated(hash = 1507368312)
    public BluetoothContact(Long id, Long timestamp, String deviceId, int rssi, int txPower,
            Boolean isUploaded) {
        this.id = id;
        this.timestamp = timestamp;
        this.deviceId = deviceId;
        this.rssi = rssi;
        this.txPower = txPower;
        this.isUploaded = isUploaded;
    }


    @Generated(hash = 352376836)
    public BluetoothContact() {
    }


    public Long getId() {
        return this.id;
    }


    public void setId(Long id) {
        this.id = id;
    }


    public Long getTimestamp() {
        return this.timestamp;
    }


    public void setTimestamp(Long timestamp) {
        this.timestamp = timestamp;
    }


    public String getDeviceId() {
        return this.deviceId;
    }


    public void setDeviceId(String deviceId) {
        this.deviceId = deviceId;
    }


    public int getRssi() {
        return this.rssi;
    }


    public void setRssi(int rssi) {
        this.rssi = rssi;
    }


    public int getTxPower() {
        return this.txPower;
    }


    public void setTxPower(int txPower) {
        this.txPower = txPower;
    }


    public Boolean getIsUploaded() {
        return this.isUploaded;
    }


    public void setIsUploaded(Boolean isUploaded) {
        this.isUploaded = isUploaded;
    }

}

File: app/src/main/java/no/simula/corona/data/model/BluetoothContact.kt
package no.simula.corona.data.model

import no.simula.corona.Utils
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

data class BluetoothContact(
    val timestamp: Long,
    val deviceId: String,
    val rssi: Int,
    val txPower: Int,
    val isUploaded: Boolean = false
) {

    companion object {

        fun fromEntity(entity: no.simula.corona.data.legacy.BluetoothContact): BluetoothContact {
            var model = BluetoothContact(
                entity.timestamp,
                entity.deviceId,
                entity.rssi,
                entity.txPower,
                entity.isUploaded
            )

            model.id = entity.id?.toLong()

            return model
        }


        fun fromEntity(entity: no.simula.corona.data.greendao.BluetoothContact): BluetoothContact {
            var model = BluetoothContact(
                entity.timestamp,
                entity.deviceId,
                entity.rssi,
                entity.txPower,
                entity.isUploaded
            )

            model.id = entity.id

            return model
        }

        fun toEntity(bluetoothContact: BluetoothContact): no.simula.corona.data.legacy.BluetoothContact {
            return no.simula.corona.data.legacy.BluetoothContact(
                bluetoothContact.timestamp,
                bluetoothContact.deviceId,
                bluetoothContact.rssi,
                bluetoothContact.txPower,
                bluetoothContact.isUploaded
            )
        }

        fun fromEntity(measurements: List<no.simula.corona.data.legacy.BluetoothContact>): List<BluetoothContact> {
            var mapped = ArrayList<BluetoothContact>()

            measurements.map {
                mapped.add(BluetoothContact.fromEntity(it))
            }

            return mapped
        }

        fun fromEntityGreenDao(models: List<no.simula.corona.data.greendao.BluetoothContact>): List<BluetoothContact> {

            var mapped = ArrayList<BluetoothContact>()

            models.map {
                mapped.add(BluetoothContact.fromEntity(it))
            }

            return mapped

        }

        fun toEntityGreenDao(model: BluetoothContact): no.simula.corona.data.greendao.BluetoothContact {

            return no.simula.corona.data.greendao.BluetoothContact(
                null, model.timestamp,
                model.deviceId,
                model.rssi,
                model.txPower,
                model.isUploaded
            )

        }

        fun toEntityGreenDao(models: List<BluetoothContact>): List<no.simula.corona.data.greendao.BluetoothContact> {

            var mapped = ArrayList<no.simula.corona.data.greendao.BluetoothContact>()

            models.map {
                mapped.add(toEntityGreenDao(it))
            }

            return mapped

        }
    }

    var id: Long? = null

    fun toJson(): JSONObject {
        val df = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        df.timeZone = TimeZone.getTimeZone("UTC")

        val obj = JSONObject()
        obj.put("time", df.format(Utils.toEpochTime(timestamp)))
        obj.put("deviceId", deviceId)
        obj.put("rssi", rssi)
        obj.put("txPower", txPower)

        return obj
    }
}

class BluetoothContactBuilder(private val timestamp: Long) {
    private var deviceId = ""
    private var rssi = 0
    private var txPower = 0


    fun setCloseContact(contactId: String, contactRssi: Int, contactTx: Int) {
        deviceId = contactId
        rssi = contactRssi
        txPower = contactTx

    }

    fun toEntity(): BluetoothContact {
        return BluetoothContact(
            timestamp,
            deviceId, rssi, txPower
        )
    }
}